src/sys/signal.rs

```diff
@@ -410,6 +410,51 @@ pub unsafe fn sigaction(signal: Signal, sigaction: &SigAction) -> Result<SigActi
     Errno::result(res).map(|_| SigAction { sigaction: oldact })
 }
 
+/// Signal management (see [signal(3p)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/signal.html))
+///
+/// Installs `handler` for the given `signal`, returning the previous signal handler if it was
+/// `SigHandler::SigDfl` or `SigHandler::SigIgn`, `None` otherwise.
+///
+/// # Examples
+///
+/// ```no_run
+/// use nix::sys::signal::{Signal, SigHandler, signal};
+///
+/// unsafe {
+///     signal(Signal::SIGINT, SigHandler::SigIgn).unwrap();
+///     assert_eq!(signal(Signal::SIGINT, SigHandler::SigDfl).unwrap(), Some(SigHandler::SigIgn));
+/// }
+/// ```
+///
+/// # Errors
+/// Returns any error from `libc::signal`, such as when setting a signal handler for
+/// `SIGKILL` or `SIGSTOP`.
+///
+/// ```no_run
+/// use nix::sys::signal::{Signal, SigHandler, signal};
+///
+/// unsafe {
+///     assert!(signal(Signal::SIGKILL, SigHandler::SigIgn).is_err());
+///     assert!(signal(Signal::SIGSTOP, SigHandler::SigIgn).is_err());
+/// }
+/// ```
+pub unsafe fn signal(signal: Signal, handler: SigHandler) -> Result<Option<SigHandler>> {
+    let signal = signal as libc::c_int;
+    let res = match handler {
+        SigHandler::SigDfl => libc::signal(signal, libc::SIG_DFL),
+        SigHandler::SigIgn => libc::signal(signal, libc::SIG_IGN),
+        SigHandler::Handler(handler) => libc::signal(signal, handler as libc::sighandler_t),
+        SigHandler::SigAction(handler) => libc::signal(signal, handler as libc::sighandler_t),
+    };
+    Errno::result(res).map(|oldhandler| {
+        match oldhandler {
+            libc::SIG_DFL => Some(SigHandler::SigDfl),
+            libc::SIG_IGN => Some(SigHandler::SigIgn),
+            _ => None, // cannot safely convert custom handler to SigHandler
```

Comment:
any suggestions here? I don't think we can get the type information of the pointer to know if it was a `SigHandler::Handler` function or a `SigHandler::SigAction` function. Another option is to return `libc::sighandler_t` here, but my thinking was that callers who needed that level of detail could just work with the  raw `libc::signal` API instead.

Reply:
I think we should always treat it like a `SigHandler::Handler`, because that should be the only type of handler that can be set with `signal(2)`.  If the program is setting the handler with `sigaction`, then it should also use `sigaction` to check the old handler.